Add support for require/define with unknown arg types at build time

chuckdumont opened the pull request:
Will transform
```javascript
  require(getDeps(), myFactoryFunction);
```
to
```javascript
(function () {
  var maybeFunction = myFactoryFunction;
  var amdDeps = getDeps();
  if (Array.isArray(amdDeps) {
    return require(amdDeps);
  }
  if (typeof maybeFunction !== "function") {
    maybeFunction = function () {};
  }
  maybeFunction.apply(void 0, amdDeps.map(function (dep) {
    return {
      require: require,
      module: module,
      exports: module.exports
    }[dep] || require(dep);
  }));
}).apply(this);
```
If either the dependency list is known to be an array, or the factory is known to be a function, at build time then the associated runtime type checking for the argument is omitted from the generated code.

define calls are transformed in a similar manner, but include code for assigning the value returned by the factory function to `module.exports`.

```javascript
(function () {
  var maybeFunction = factory;
  var amdDeps = deps;
  if (typeof amdDeps === 'string') {
    amdDeps = ['require', 'exports', 'module'];
  }
  if (typeof maybeFunction !== "function") {
    var amdFactoryResult = maybeFunction;
    maybeFunction = function () {
      return amdFactoryResult;
    };
  }
  var amdDefineResult = maybeFunction.apply(void 0, amdDeps.map(function (dep) {
    return {
      require: require,
      module: module,
      exports: module.exports
    }[dep] || require(dep);
  }));
  typeof amdDefineResult !== "undefined" && (module.exports = amdDefineResult);
}).apply(this);
```

msrose commented:
Hey @chuckdumont thanks very much for the pull request, I've approved builds to start running and will take a closer look in the next week

chuckdumont commented:
Note that as implemented, you cannot pass a JavaScript Arguments object to require.  That's because identifying an Arguments object is not that simple without a library utility (see [here](https://stackoverflow.com/questions/7656280/how-do-i-check-whether-an-object-is-an-arguments-object-in-javascript)) and I didn't want to bloat the generated code any further.  If this functionality is requested by anyone, then it could perhaps be enabled by an option setting.

msrose commented:
Hi again @chuckdumont -- nice work on this, I've had a closer look and have some high level thoughts:

1. Very clever implementation for handling "maybe array" dependency lists, I'd like to purpose it for `define` as well as part of addressing the next two points:
2. The plugin already has handling for "maybe function" factories for `define`, see [here in the code](https://github.com/msrose/babel-plugin-transform-amd-to-commonjs/blob/master/src/helpers.js#L153). I'd like to follow this same code path for require so the codegen is consistent and handling of AMD keywords is guaranteed to be consistent as the plugin evolves going forward. Please modify your solution to also use this code path for `require` factories. The main difference is that **require does not need a check on the return value of the factory** as require isn't used to define modules, so we can skip `createModuleExportsResultCheck` when using the existing code for require.
3. At the same time, we can modify that function in order to get both `define` and `require` handling for "maybe array" dependency lists -- your existing solution should serve that purpose nicely :)

Please let me know if you have any questions or thoughts around the requested changes. Once the changes get posted we can discuss again and I'll leave some more detailed comments on the implementation itself to discuss code organization/naming/edge cases/etc.

chuckdumont commented:
I figured the `requre` and `define` "maybe function" handling could be combined, but wasn't sure if you'd want to do that with this PR or in a follow-on PR.  I'm happy to make the change you suggest.

chuckdumont commented:
@mrose - I've made the updates you requested.  Please re-review.  The README still needs updating wanted to make sure that I'm on the right track.  Thanks.

chuckdumont commented:
@msrose - please advise if the current changes are what you're looking for.

msrose commented:
Hey @chuckdumont sorry for the delay, haven't had a chance to take a close look just yet. ETA by the end of this week.

chuckdumont commented:
Hold off on reviewing this.  The support for define needs more work.

msrose commented:
Thanks for the note, I started going over a few parts the other day and just testing some of the more bizarre edge cases with requirejs to verify we're replicating the behaviour correctly - thanks for your hard work on this so far, lots of tricky cases to handle :)

chuckdumont commented:
@msrose - I've updated the PR to correctly handle the define calls where it is ambiguous whether the first argument is a module name or a dependency list.  The generated code now checks for this case and reacts appropriately.

chuckdumont commented:
I believe I'm handling all the edge cases now.  Please review.

msrose commented:
Thanks @chuckdumont, this is on my todo list for this weekend. Thanks for your patience.

msrose commented on src/index.js:
```diff
@@ -63,13 +64,28 @@ module.exports = ({ types: t }) => {
+          node.expression.arguments.length
+        )
+      );
+      return;
```
This early return ultimately short-circuits the existing code and I believe there are some dead code paths further down that won't be hit any longer -- I'd prefer if we could use the existing code paths and add the additional edge cases in there, since the code is already well tested and been used in the published package for several years now, but I understand why it's easier from an implementation perspective to do it this way. Let's leave it for now, but please create an issue linking here and assign it to me, so that there is a backlog item for cleaning up the duplicate/dead code paths, as I don't want to increase the scope of this PR any further.

chuckdumont replied:
I've tweaked the conditional for entering the new code so that all the existing use cases still use the old code.  All the previously existing unit tests still run unmodified and the new unit tests only test cases that weren't previously handled.

msrose replied:
Thanks, I think this is a smart call just for extra safety 👍

msrose commented on tests/require.test.js:
```diff
@@ -124,16 +125,6 @@ describe('Plugin for require blocks', () => {
     `);
   });
 
-  it('ignores non-function factories', () => {
```
Hmm... why is this test no longer valid?

chuckdumont replied:
Restored

msrose commented on tests/test-helpers.js:
```diff
@@ -36,7 +36,76 @@ const checkMaybeFunction = (factory, dependencies, identifier = MAYBE_FUNCTION)
     `;
 };
 
+const checkVarArgsResult = (
```
Just to make this function name more descriptive, can we please change to `checkVariableDepAndFactoryResult` or similar.

Also similar to a previous comment, let's make the arguments an object instead, i.e. `myFunc({ factory, dependencies, ... })`, for increased readability at call sites and ease of future refactors.

msrose commented on src/helpers.js:
```diff
@@ -195,6 +204,230 @@ module.exports = ({ types: t }) => {
+         * 'deps.map(<amdFactoryArgsMapper>)''
+         */
+        t.callExpression(t.memberExpression(depsIdentifier, t.identifier('map')), [
+          getAmdFactoryArgsMapper(path, opts),
```
`path` and `opts` are not parameters to this function, so let's remove.

chuckdumont replied:
Done

msrose reviewed (changes requested):
Nice work so far 🎉  let me know if you have any questions regarding the comments.

chuckdumont commented on src/helpers.js:
```diff
@@ -195,6 +204,230 @@ module.exports = ({ types: t }) => {
+        // if (typeof amdDeps === 'string') {
+        //   amdDeps = ['require', 'exports', 'module'];
+        // }
+        if (arity === 2) {
```
It needs to be a separate conditional because of the else clause on the `if (isDefineCall)`.  If the conditionals are combined, as you suggest, then it changes the logic and we'd need to change the `else` to `else if (!isDefineCall)` which I don't think is any better.

chuckdumont commented on src/helpers.js:
```diff
@@ -195,6 +204,230 @@ module.exports = ({ types: t }) => {
+    }
+    // Invoke the factory function.
+    const factoryInvocation = t.callExpression(
+      t.memberExpression(factoryIdentifier, t.identifier('apply')),
```
`.apply` is used so that we can pass the `this` reference.  We could use an arrow function instead, but I didn't think that arrow functions should be used in generated code.  It's necessary to preserve the `this` context because of the following common use pattern:

```javascript
require(['foo'], function(foo) {
  foo.saySomething(this.sayWhat);
}.bind(this));
```

If we use a normal (non-arrow) function call, then the `this` reference within the function will be the `window` object.

msrose replied:
But we're passing undefined (void 0) as the this arg every time, so isn't that just the same as calling the function without using apply?

chuckdumont commented on src/helpers.js:
```diff
@@ -195,6 +204,230 @@ module.exports = ({ types: t }) => {
+    );
+    if (isDefineCall) {
+      const resultCheckIdentifier = getUniqueIdentifier(path.scope, AMD_DEFINE_RESULT);
+      blockStatements.push.apply(
```
Not exactly the same because `createModuleExportsResultCheck` returns an array of block statements.  You're suggested change will push the array object onto `blockStatements` rather than adding the individual elements to the `blockStatements` array.  We could use `blockStatements.concat` instead, but that requires a re-assignment and I don't see a reason to prefer one approach over the other.

msrose replied:
Ah I see, let's use the spread operator to make it clearer what's happening then. (The source is transpiled before publish time so this is fine for backcompat.) e.g. `blockStatements.push(...arr)`

chuckdumont replied:
Ah, that wrapper.  I thought we were talking about the outer wrapper.  So the reason for using apply here is so that the function arguments can be passed as an array (the result of calling `amdDeps.map`).

msrose replied:
Ahh got it, thanks for the explanation

chuckdumont commented on src/helpers.js:
```diff
@@ -195,6 +204,230 @@ module.exports = ({ types: t }) => {
+
+    // If we don't know that the factory is a function, then we need to check the type
+    // at runtime.
+    if (!isFunctionExpression(factory)) {
```
This code path is now executed only when the dependency list is also not an array type.  so it makes more sense to use the new code here to check the factory type rather than try to reuse the existing code.

chuckdumont commented on src/helpers.js:
```diff
@@ -195,6 +204,230 @@ module.exports = ({ types: t }) => {
+      }
+      // Convert array-like objects to real arrays
+      // amdDeps = [].slice.call(amdDeps);
+      blockStatements.push(
```
Done

chuckdumont commented:
@msrose - I believe I addressed all of your comments except for the outer IIF.  If you still against it, I can remove it, but I'd like your feedback on my response.

chuckdumont commented:
@msrose - another reason for wrapping the generated code in an IIF is to avoid syntax errors when the require or define call is not stand-alone.  Consider the following example:

```javascript
return foo ? require(['foo'], foofactory) : require(['bar'], barfactory);
```

Granted, this is a rather contrived example but it illustrates the point.  With the generated code wrapped in the IIF the transformed code will work as expected.  Without the IIF, there will be syntax errors all over the place.

msrose commented:
Hm yeah that's an interesting case -- I'm fine to leave the IIFE for now then

msrose reviewed (approved):
Thanks again for your work on this @chuckdumont and for taking on the extra scope -- great job and thank you for addressing all the comments in detail. The only part left is a small docs update but I'm happy to merge this in its current state since you've added the examples in your PR description. 👏  I will follow up with a docs update and add some small integ tests, then will cut a release.
